### PR TITLE
Add a loss comparison script

### PR DESCRIPTION
**Pull request by fegin**
**Summary**
This PR adds a script to validate benchmark loss against the baseline, streamlining the comparison process. The script is particularly useful for users and developers who refactor code and want to ensure that their changes do not affect the original loss. Additionally, it provides a quick verification tool when introducing new parallelism or training logic. But with new parallelism or training logic, a comprehensive validation is required. Please refer to the large-scale training procedures described in docs/converging.md.

**Note: Both seed checkpoint and set_deterministic are used by default and only seed checkpoint can be disabled.**

**Example usage:**
*Compare main branch with chienchin/pp_cp_flex. This should result in same losses.*
```
scripts/loss_compare.sh main chienchin/pp_cp_flex
```

*Compare the same code with different parallelisms with seed checkpoint.*
*This serves a quick understanding of whether two parallelism maintain a similar loss curve.*
*A comprehensive validation is required to actually verify the accuracy.*
```
scripts/loss_compare.sh . . --baseline-train-options="--parallelism.data_parallel_replicate_degree=1" --benchmark-train-options="--parallelism.data_parallel_replicate_degree=2" --steps=30
```

### CODE REVIEW

File: scripts/loss_compare.sh [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+    echo "$message" >> "$STATS_FILE"
+}
+
+#=============================================================================
+# HELP AND VALIDATION FUNCTIONS
+#=============================================================================
+
+# Function to show usage
+show_usage() {
+    echo "$LOG_PREFIX Usage: $0 <baseline_commit> <benchmark_commit> [OPTIONS]"
```
Comment (tianyu-l) [inline scripts/loss_compare.sh]: "benchmark" sounds ambiguous.
Maybe "baseline vs. test", or ""main vs. test" or even "control vs. test"

File: scripts/loss_compare.sh [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+BASELINE_TRAIN_OPTIONS=""
+BENCHMARK_TRAIN_OPTIONS=""
+ENABLE_SEED_CHECKPOINT=true
+OUTPUT_FOLDER="loss_compare"
+
+# Command templates
+FIXED_OPTIONS="--training.deterministic --training.seed=42"
+SEED_CHECKPOINT_CMD="./run_train.sh --checkpoint.create_seed_checkpoint --checkpoint.enable ${FIXED_OPTIONS}"
+TRAIN_CMD_WITH_CHECKPOINT="./run_train.sh --checkpoint.enable --checkpoint.export_dtype=bfloat16 ${FIXED_OPTIONS}"
+TRAIN_CMD_WITHOUT_CHECKPOINT="./run_train.sh ${FIXED_OPTIONS}"
```
Comment (tianyu-l) [inline scripts/loss_compare.sh:48]: Shall we add a unit test for this script? I'm worried that it will be broken easily.

Comment (fegin) [reply scripts/loss_compare.sh:48]: I think a test is a valid request. But what tests are in you mind? I feel an integration test is more useful than unittest. In most cases, this script just pass the command to the run_train.sh except for some preset configurations. And then the script pass the output. So I would assume that all of these can only be verified through an integration test.

An alternative is to set this script as the minimal loss guard in the CI, like running `scripts/loss_compare.sh main PR_BRANCH` to verify the loss and this script itself at the same time.

Let me know what you think about this proposal.

Comment (tianyu-l) [reply scripts/loss_compare.sh:48]: Yes, what I meant was just an integration test which guarantees that this script is not broken by future commits.

It indeed sounds better to have deterministic test
> An alternative is to set this script as the minimal loss guard in the CI, like running scripts/loss_compare.sh main PR_BRANCH to verify the loss and this script itself at the same time.

I don't how often it will be broken by pytorch nightly changes. We can give it a try.

Comment (fegin) [reply scripts/loss_compare.sh:48]: > I don't how often it will be broken by pytorch nightly changes. We can give it a try.

We should have faith on our PyTorch, lol. But if the test is indeed broken by some PyTorch nightly, I think that gives us a good signal that our loss accuracy may be at risk.
